### PR TITLE
[SPARK-6862][Streaming][WebUI] Add BatchPage to display details of a batch

### PR DESCRIPTION
**Pull request by zsxwing**
This is an initial commit for SPARK-6862. Once SPARK-6796 is merged, I will add the links to StreamingPage so that the user can jump to BatchPage.

Screenshots:
success
failure

### CODE REVIEW

**Comment by zsxwing** [comment]
My test codes:
```Scala
import org.apache.spark._
import org.apache.spark.streaming._

object StreamingApp {

  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local[2]").setAppName("NetworkWordCount").set("spark.streaming.concurrentJobs", "3")
    val ssc = new StreamingContext(conf, Seconds(10))
    val lines = ssc.socketTextStream("localhost", 9999)
    val words = lines.flatMap(_.split(" "))
    val pairs = words.map(word => {
      Thread.sleep(1000)
      (word, 1)
    })
    val wordCounts = pairs.reduceByKey((x: Int, y: Int) => x + y, 2)
    wordCounts.foreachRDD {
      rdd => rdd.foreach { v =>
        println(v)
      }
    }
    wordCounts.foreachRDD { rdd =>
      rdd.foreach { v =>
        println(v)
      }
      rdd.foreach { v =>
        println(v)
        //throw new RuntimeException("Oops!")
      }
    }
    ssc.start()
    ssc.awaitTermination()
  }
}
```

**Comment by zsxwing** [comment]
cc @tdas

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala [hunk 1]
```diff
@@ -58,4 +58,11 @@ case class BatchInfo(
    */
   def totalDelay: Option[Long] = schedulingDelay.zip(processingDelay)
     .map(x => x._1 + x._2).headOption
+
+  /**
+   * The number of recorders received by the receivers in this batch.
+   */
+  def numRecords: Long = receivedBlockInfo.map { case (_, infos) =>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala]: +1 on exposing this.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 2]
```diff
@@ -170,8 +170,10 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
     ssc.waiter.notifyError(e)
   }
 
-  private class JobHandler(job: Job) extends Runnable {
+  private class JobHandler(job: Job) extends Runnable with Logging {
     def run() {
+      ssc.sc.setLocalProperty(JobScheduler.BATCH_TIME_PROPERTY_KEY, job.time.milliseconds.toString)
+      ssc.sc.setLocalProperty(JobScheduler.OUTPUT_OP_ID_PROPERTY_KEY, job.outputOpId.toString)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: This setting and clearing of property needs to be done in a `try....finally` so that it gets cleared in case of any exception.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 3]
```diff
@@ -179,7 +181,14 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
...
       }
+      ssc.sc.setLocalProperty(JobScheduler.BATCH_TIME_PROPERTY_KEY, null)
+      ssc.sc.setLocalProperty(JobScheduler.OUTPUT_OP_ID_PROPERTY_KEY, null)
       eventActor ! JobCompleted(job)
     }
   }
 }
+
+private[streaming] object JobScheduler {
+  private[streaming] val BATCH_TIME_PROPERTY_KEY = "spark.streaming.internal.batchTime"
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: If the object private[streaming] then all the field are already private[streaming], isnt it?

**Comment by tdas** [comment]
Its looking pretty good! Here are some preliminary comments. I will do a more detailed pass tomorrow during the day. 

1. There should be tests in the StreamingListenerSuite that tests the local property setting. 
2. In the screenshots, is there any way to limit the max col size of the error? When there is error, it squishes all the other columns.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+  private def columns: Seq[Node] = {
+    <th>Output Op Id</th>
+      <th>Description</th>
+      <th>Duration</th>
+      <th>Job Id</th>
+      <th>Duration</th>
+      <th class="sorttable_nosort">Stages: Succeeded/Total</th>
+      <th class="sorttable_nosort">Tasks (for all stages): Succeeded/Total</th>
+      <th>Last Error</th>
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: Why is this `Last Error` and not just `Error`

**Comment by tdas** [comment]
@zsxwing Any thoughts on this?

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala (hunk 4 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: I was thinking `Last Error` of tasks. But here is Stage. I agree it should be `Error`. Fixed.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala (hunk 2 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: Put them to `finally`.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala (hunk 3 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: Right. I just removed `private[streaming]`.

**Comment by zsxwing** [comment]
> There should be tests in the StreamingListenerSuite that tests the local property setting.

As I explained in `https://github.com/apache/spark/pull/5288#issuecomment-92579899`, we may get wrong properties in `SparkListenerJobStart.properties`. Because the tests will be non-deterministic, I think it's better to add them after @JoshRosen fixes the issue.

> In the screenshots, is there any way to limit the max col size of the error? When there is error, it squishes all the other columns.

Let me try some css style to fix it.

**Comment by tdas** [comment]
Why would we get wrong properties in this case? We are explicitly setting it in the thread that is launching the Spark job, so there is no question of inheriting local properties by child threads, etc. So that issue should not affect this case, isnt it? Or am I missing something?

**Comment by zsxwing** [comment]
> Why would we get wrong properties in this case? We are explicitly setting it in the thread that is launching the Spark job, so there is no question of inheriting local properties by child threads, etc. So that issue should not affect this case, isnt it? Or am I missing something?

Without #5288, SparkListenerJobStart is created like `SparkListenerJobStart(..., localProperties.get)`. So the `properties` in `SparkListenerJobStart` is the same object of `localProperties.get`, which is supposed to be limited in its thread. 

And after `job.run`, we will clear some properties in `localProperties.get`, because `SparkListenerJobStart.properties` is the same object of `localProperties.get`, if `job.run` is very fast, we may not find the properties we need, because they have been cleared.

**Comment by zsxwing** [comment]
I set the max-width of the error column to `300px`. Here is the new screenshot:

batch_page

**Comment by zsxwing** [comment]
Please ignore that `Processing time` and `Total delay` are `-`. It's an issue about `JobScheduler`. I will open a JIRA for it.

**Comment by zsxwing** [comment]
> Please ignore that Processing time and Total delay are -. It's an issue about JobScheduler. I will open a JIRA for it.

After thinking it carefully, this may not be an issue, since StreamingContext should be stopped in such case. @tdas, do you think if we still need to show Processing time and Total delay in this case?

**Comment by zsxwing** [comment]
@tdas this is ready for code review.

**Comment by zsxwing** [comment]
In addition, because of `rowspan`, the table cannot be sortable.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala [hunk 5]
```diff
@@ -25,15 +25,43 @@ import scala.util.Try
...
+    }
+    _outputOpId
+  }
+
+  def setOutputOpId(outputOpId: Int) {
+    if (isSet) {
+      throw new IllegalStateException("Cannot call setOutputOpId more than once")
+    }
+    isSet = true
+    _id = "streaming job " + time + "." + outputOpId
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: use string interpolation.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala [hunk 6]
```diff
@@ -25,15 +25,43 @@ import scala.util.Try
  */
 private[streaming]
 class Job(val time: Time, func: () => _) {
-  var id: String = _
-  var result: Try[_] = null
+  private var _id: String = _
+  private var _outputOpId: Int = _
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: Could you define what is `id` and `outputId` (especially that output id is unique within the jobset)

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 7]
```diff
@@ -40,6 +45,8 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
   private var totalProcessedRecords = 0L
   private val receiverInfos = new HashMap[Int, ReceiverInfo]
 
+  private val batchTimeToJobIds = new HashMap[Time, ArrayBuffer[(OutputOpId, JobId)]]
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Isnt it more intuitive to use a HashMap than an ArrayBuffer[(key, value)]?

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Aaah, I see there can be multiple JobIds for the same OutputId.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+      </table>
+    }
+  }
+
+  def render(request: HttpServletRequest): Seq[Node] = {
+    val batchTime = Option(request.getParameter("id")).map(id => Time(id.toLong)).getOrElse {
+      throw new IllegalArgumentException(s"Missing id parameter")
+    }
+    val formattedBatchTime = UIUtils.formatDate(batchTime.milliseconds)
+    val (batchInfo, jobInfos) = streaminglistener.synchronized {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: I think its quite awkward and unintuitive that a code outside StreamingJobProgressListener is synchronizing on the same object as the internals of StreamingJobProgressListener. Its best to move this code inside to atomically get everything.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+      val lastStageName = lastStageInfo.map(_.name).getOrElse("(Unknown Stage Name)")
+      val lastStageDescription = lastStageData.flatMap(_.description).getOrElse("")
+      val duration: Option[Long] = {
+        job.submissionTime.map { start =>
+          val end = job.completionTime.getOrElse(System.currentTimeMillis())
+          end - start
+        }
+      }
+      val lastFailureReason = job.stageIds.sorted.reverse.flatMap(sparkListener.stageIdToInfo.get).
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: nit: this is a lot of chaining and hard to follow. please break it down.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+      sparkListener.activeJobs.get(jobId).orElse {
+        sparkListener.completedJobs.find(_.jobId == jobId).orElse {
+          sparkListener.failedJobs.find(_.jobId == jobId)
+        }
+      }
+    }
+
+    // Group jobInfos by OutputOpId firstly, then sort them.
+    // E.g., [(0, 1), (1, 3), (0, 2), (1, 4)] => [(0, [1, 2]), (1, [3, 4])]
+    val outputOpIdWithJobIds: Seq[(OutputOpId, Seq[JobId])] =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: put this multi-line into a `{...}`

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 7 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: `job` in `batchTimeToJobIds ` is confusing as there is a streaming Job, and a spark job. Please call it `sparkJob` wherever you are referring to spark jobs. In fact, call the type `JobId` as `SparkJobId`.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+    })
+    val formattedOutputOpDuration =
+      if (jobDurations.exists(_ == None)) {
+        // If any job does not finish, set "formattedOutputOpDuration" to "-"
+        "-"
+      } else {
+        UIUtils.formatDuration(jobDurations.flatMap(x => x).sum)
+      }
+
+    def makeJobRow(job: JobUIData, isFirstRow: Boolean): Seq[Node] = {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: The nesting too deep in this inner function. Its really hard to follow the control flow. Please move this `makeJobRow` outside the `makeOutputOptIdRow` and add scala docs.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+          end - start
+        }
+      }
+      val lastFailureReason = job.stageIds.sorted.reverse.flatMap(sparkListener.stageIdToInfo.get).
+        dropWhile(_.failureReason == None).take(1). // get the first info that contains failure
+        flatMap(info => info.failureReason).headOption.getOrElse("")
+      val formattedDuration = duration.map(d => UIUtils.formatDuration(d)).getOrElse("-")
+      val detailUrl = s"${UIUtils.prependBaseUri(parent.basePath)}/jobs/job?id=${job.jobId}"
+      <tr>
+        {if(isFirstRow) {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: The formatting is weird here and really hard to follow. I cant even figure out where the `{` of the if condition ends. Please indent properly.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+          {formattedDuration}
+        </td>
+        <td class="stage-progress-cell">
+          {job.completedStageIndices.size}/{job.stageIds.size - job.numSkippedStages}
+          {if (job.numFailedStages > 0) s"(${job.numFailedStages} failed)"}
+          {if (job.numSkippedStages > 0) s"(${job.numSkippedStages} skipped)"}
+        </td>
+        <td class="progress-cell">
+          {UIUtils.makeProgressBar(started = job.numActiveTasks, completed = job.numCompletedTasks,
+          failed = job.numFailedTasks, skipped = job.numSkippedTasks,
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: indent scala code correctly.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala (hunk 8 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: Also, this is not job infos! This is only the outputId to jobId mapping. That was very confusing as in the Spark UI stuff there is StageInfo, JobUIData, StageUIData, and all sorts of classes which actually contain info.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 14]
```diff
@@ -180,4 +205,21 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
...
+    }
+  }
+
+  def getJobInfos(batchTime: Time): Option[Seq[(OutputOpId, JobId)]] = synchronized {
+    batchTimeToJobIds.get(batchTime).map(_.toList)
+  }
+}
+
+private[streaming] object StreamingJobProgressListener {
+  type JobId = Int
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: As I said at other places, `JobId` is confusing its not clear whether it is a streaming job or a spark job. Please call it `SparkJobId`. And also please rename all the places you are passing spark's job id as `sparkJobId`.

**Comment by tdas** [comment]
Overall, I get the code. However it is hard to read because of the following.

1. confusion on the meaning of "job", which could be streaming job or spark job. Please make sure to refer all the places in streaming code where you refer to spark job are named as `sparkJob`. This includes ids, infos, etc. in the StreamingJobProgressListener and BatchPage.

2. This tuple (OutputOpId, JobId) is being passed around. It maybe cleaner to create something like a `BatchUIData` object (similar to `JobUIData`, `StageUIData`, `TaskUIData` in `JobProgressListener`). That will encapsulate all the batch and output op related data inside a single class, as opposed to being spread across multiple data structures. So `BatchUIData` will have outputId --> jobIds inside it. I think that is more intuitive as the output id and jobids are automatically scoped within the BatchUIData. This is safer as well for cleanup; when the `BatchUIData` object is cleared, all associated information is cleared.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala (hunk 6 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/Job.scala]: Added comments on the `def outputOpId: Int` and `def id: String`.

**Comment by zsxwing** [comment]
All comments addressed. Thanks a lot for reviewing this one.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+
+import org.apache.spark.streaming.Time
+import org.apache.spark.streaming.scheduler.BatchInfo
+import org.apache.spark.ui.{UIUtils, WebUIPage}
+import org.apache.spark.streaming.ui.StreamingJobProgressListener.{SparkJobId, OutputOpId}
+import org.apache.spark.ui.jobs.UIData.JobUIData
+
+private[ui] case class BatchUIData(
+    var batchInfo: BatchInfo = null,
+    outputOpIdToSparkJobIds: Map[OutputOpId, ArrayBuffer[SparkJobId]] = Map()) {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: Why is this a case class? the first field is a var and the second field does not require specification in the constructor.

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchPage.scala]: Also since this is used by StreamingJobProgressListener and then by BatchPage, this class should be either in a separate class, or at least in the StreamingJobProgressListener class.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 16]
```diff
@@ -40,6 +43,8 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
   private var totalProcessedRecords = 0L
   private val receiverInfos = new HashMap[Int, ReceiverInfo]
 
+  private val batchTimeToBatchUIData = new HashMap[Time, BatchUIData]
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: This is good! But still not enough. The whole point of adding BatchUIData to encapsulate BatchInfo and outputOpId --> SparkJobId in it. So that all of that can be cleaned easily and we dont have to maintain separate data structures of BatchInfos (that is, waitingBatchInfos, runningBatchInfos, etc) and BatchUIData. So we should convert all (waiting/running/completed)BatchInfos to ***BatchUIData. Then we dont need to add this extra `batchTimeToBatchUIData` any more. And we dont need two methods - getBatchInfo and getBatchUIData, only the latter is needed.

**Comment by tdas** [comment]
You have to add tests to the StreamingJobProgressListener to test the use thread-local properties to combine streaming and spark event information. This is very subtle code and can silently break if someone change the behavior of thread-local properties in spark job properties. Also test the expected number of output ids, and Spark job ids using a program with multiple output ops and multiple Spark actions/jobs in each output op. 

You can test these in the  StreamingJobProgressListenerSuite in the same way you test the batch and receiver events - by adding spark events in the mix.

**Comment by zsxwing** [comment]
@tdas, I have a question about the failure case: If a Spark job throws an exception, it will propagate to `ssc.awaitTermination()`. In such case, I suppose the user will call `ssc.stop()` and exit the application. So the user cannot have a chance to view the information in the `Last Error` column. Right? If so, it's weird that we have a `Last Error` column. (For my test, I just comment out `ssc.stop()` so that I can see the exception in the UI)

**Comment by tdas** [comment]
That's a good point. But if someone actually catches the exception within foreachRDD and moves on, then it is worth showing the exception. But if the exception is not caught, any kind of exception is forwarded to awaitTermination and the streamingContext is automatically stopped.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 16 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: > So we should convert all (waiting/running/completed)BatchInfos to ***BatchUIData. Then we dont need to add this extra batchTimeToBatchUIData any more.

It's a bit complex here. Because `onJobStart` and `onBatchXXX` messages are processed in different threads, we may not be able to get the corresponding BatchInfo in `onJobStart`. In such case, if we convert all `(waiting/running/completed)BatchInfos` to `***BatchUIData`, we still need to have an extra data structure to store `BatchUIData`s that does not have a corresponding `BatchInfo`.

**Comment by zsxwing** [comment]
retest this please

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 17]
```diff
@@ -40,6 +47,30 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
   private var totalProcessedRecords = 0L
   private val receiverInfos = new HashMap[Int, ReceiverInfo]
 
+  // Because onJobStart and onBatchXXX messages are processed in different threads,
+  // we may not be able to get the corresponding BatchInfo when receiving onJobStart. So here we
+  // cannot use a map of (Time, BatchUIData).
+  private[ui] val batchTimeToOutputOpIdSparkJobIdPair =
+    new LinkedHashMap[Time, ArrayBuffer[(OutputOpId, SparkJobId)]] {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Ideally, I would like this `(OutputOpId, SparkJobId)` to be a case class since it occurs multiple time. But I am okay without it.

**Comment by tdas** [comment]
High level points as discussed offline.

We are planning to increase the number of retainedBatches, hence we need to be careful about all the data that we retain in memory. And its not a good idea to retain BatchInfo objects in memory because it contains a lot of arbitrary `ReceivedBlockInfo` objects, which can have arbitrary metadata ( #5732 ). So a better design would be to train only `BatchUIData` objects which contains only the data necessary for rendering the UI (timings, numrecords, etc.). So `waitingBatches`, `runningBatches`, `completedBatches` should have only `BatchUIData` objects.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 17 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Actually, since there are other changes to be made, its a good idea to do this as well.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 18]
```diff
@@ -17,29 +17,58 @@
...
 
 private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
-  extends StreamingListener {
+  extends StreamingListener with SparkListener {
 
-  private val waitingBatchInfos = new HashMap[Time, BatchInfo]
-  private val runningBatchInfos = new HashMap[Time, BatchInfo]
-  private val completedBatchInfos = new Queue[BatchInfo]
-  private val batchInfoLimit = ssc.conf.getInt("spark.streaming.ui.retainedBatches", 100)
+  private val waitingBatchUIDatas = new HashMap[Time, BatchUIData]
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: `Datas` sounds really weird. Just leave it as `BatchUIData`. Or if you want to follow other hashmaps naming styles `timeToWaitingBatchUIData`.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/BatchUIData.scala [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+  def totalDelay: Option[Long] = processingEndTime.map(_ - submissionTime)
+
+  /**
+   * The number of recorders received by the receivers in this batch.
+   */
+  def numRecords: Long = receiverNumRecords.map(_._2).sum
+
+  def canEqual(other: Any): Boolean = other.isInstanceOf[BatchUIData]
+
+  override def equals(other: Any): Boolean = other match {
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/BatchUIData.scala]: What is `equals` and `hashCode` useful for? Unit tests? If so, then could this be a case class?

File: streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala [hunk 20]
```diff
@@ -86,6 +108,63 @@ class UISeleniumSuite
...
+
+        // Check stage progress
+        findAll(cssSelector(""".stage-progress-cell""")).map(_.text).toSeq should be
+          (List("1/1", "1/1", "1/1", "0/1 (1 failed)"))
+
+        // Check job progress
+        findAll(cssSelector(""".progress-cell""")).map(_.text).toSeq should be
+          (List("1/1", "1/1", "1/1", "0/1 (1 failed)"))
+
+        // Check stacktrack
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/UISeleniumSuite.scala]: stacktrack --> stacktrace

File: streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala [hunk 21]
```diff
@@ -64,13 +68,64 @@ class StreamingJobProgressListenerSuite extends TestSuiteBase with Matchers {
...
+    properties2.setProperty(JobScheduler.BATCH_TIME_PROPERTY_KEY, Time(1000).milliseconds.toString)
+    properties2.setProperty(JobScheduler.OUTPUT_OP_ID_PROPERTY_KEY, 0.toString)
+    val jobStart2 = SparkListenerJobStart(jobId = 1,
+      0L, // unused
+      Nil, // unused
+      properties2)
+    listener.onJobStart(jobStart2)
+
+    val properties3 = new Properties()
+    properties3.setProperty(JobScheduler.BATCH_TIME_PROPERTY_KEY, Time(1000).milliseconds.toString)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: Please put these two lines for setting the properties in a function. Will reduce duplicate code a lot.

File: streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala [hunk 22]
```diff
@@ -64,13 +68,64 @@ class StreamingJobProgressListenerSuite extends TestSuiteBase with Matchers {
...
+    val jobStart3 = SparkListenerJobStart(jobId = 0,
+      0L, // unused
+      Nil, // unused
+      properties3)
+    listener.onJobStart(jobStart3)
+
+    val properties4 = new Properties()
+    properties4.setProperty(JobScheduler.BATCH_TIME_PROPERTY_KEY, Time(1000).milliseconds.toString)
+    properties4.setProperty(JobScheduler.OUTPUT_OP_ID_PROPERTY_KEY, 1.toString)
+    val jobStart4 = SparkListenerJobStart(jobId = 1,
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: In fact, this code to create `SparkListenerJobStart` is also repeated. Please put them in a function.

File: streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala [hunk 23]
```diff
@@ -116,4 +171,73 @@ class StreamingJobProgressListenerSuite extends TestSuiteBase with Matchers {
     listener.retainedCompletedBatches.size should be (limit)
     listener.numTotalCompletedBatches should be(limit + 10)
   }
+
+  test("disorder onJobStart and onBatchXXX") {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: disorder --> out-of-order

File: streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala [hunk 24]
```diff
@@ -116,4 +171,73 @@ class StreamingJobProgressListenerSuite extends TestSuiteBase with Matchers {
     listener.retainedCompletedBatches.size should be (limit)
     listener.numTotalCompletedBatches should be(limit + 10)
   }
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: Is there a test for testing cleanup of all the data structures in StreamingJobProgressListener?

Comment (zsxwing) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: What's the meaning of "cleanup"? I cannot find any relating method in `StreamingJobProgressListener`.

**Comment by zsxwing** [comment]
@tdas `UISeleniumSuite` is unstable because we may get a wrong batchTime and outputOpId properties due to #5288.

**Comment by tdas** [comment]
I have merged #5288. Let me take another look at this PR.

**Comment by tdas** [comment]
retest this please.

File: streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala (hunk 24 above)
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: I meant that all the data structures are cleaned up based on the retained batches limit. I just want to ensure that there are no memory leaks.

Comment (zsxwing) [inline streaming/src/test/scala/org/apache/spark/streaming/ui/StreamingJobProgressListenerSuite.scala]: Yes. The limitation of the `batchTimeToOutputOpIdSparkJobIdPair`'s size is at the end of `test("out-of-order onJobStart and onBatchXXX")`.

**Comment by tdas** [comment]
rest this please.

**Comment by tdas** [comment]
retest this please.

**Comment by zsxwing** [comment]
retest this please.

**Comment by tdas** [comment]
The test has passed. Merging this! Thanks @zsxwing !
